### ISSUE TITLE
[Keras 3 OpenVINO Backend] Improve operator coverage and functional parity

### ISSUE THREAD

**Issue by goyaladitya05** [openvinotoolkit/openvino#34017]
# Keras 3 OpenVINO Backend Functional Parity Roadmap

This issue tracks the implementation of missing operators across the entire OpenVINO backend suite for Keras 3.  
The operations are implemented in `keras/src/backend/openvino`.

### 1. Numpy (numpy.py)

- [x] #34255
- [x] #34021
- [x] #34025
- [x] #34242
- [x] #34225
- [x] #34054
- [x] #34055
- [x] #34056
- [x] #34558
- [x] #34559
- [x] #34254
- [x] #34220
- [x] #34129
- [x] #34163
- [x] #34128
- [x] #34112
- [x] #34130
- [x] #34050
- [x] #34218
- [x] #34217
- [x] #34214
- [x] #34166
- [x] #34216
- [x] #34256
- [x] #34670 
- [x] #34701 
- [x] #34409
- [x] #34867 
- [x] #35267 
- [x] #35506

#### Bitwise Operations
- [x] #34033
- [x] #34035
- [x] #34036

#### NaN-Aware Operations
- [x] #34139
- [x] #34226
- [x] #34401
- [x] #34557
- [x] #34921 
- [x] #35502 

#### Complex Numbers
- [x] #34106
- [x] #34238
- [x] #34034

#### Window Functions
- [x] #34156
- [x] #34197

### 2. Random (random.py)

- [x] #34243
- [x] #34251
- [x] #34250
- [x] #34263
- [x] #34308

### 3. Math (math.py)

- [x] #34245
- [x] #34276
- [x] #34418
- [x] #34329
- [x] #34412
- [x] #35312 

### 4. Linalg (linalg.py)

- [x] #34269
- [x] #34127
- [x] #34296
- [x] #34510
- [x] #34520
- [x] #34540 
- [x] #34793 
- [x] #34825 
- [x] #35002 
- [x] #35501 

### 5. Neural Networks (nn.py)

- [x] #34248
- [x] #34274
- [x] #34275
- [x] #34417
- [x] #34522
- [x] #34523
- [x] #35181
- [x] #35264 
- [x] #35585

### 6. Image (image.py)

- [x] #34377
- [x] #34795 
- [x] #34796 
- [x] #35208 
- [x] #35412
- [x] #35584  

### 7. RNN (rnn.py)

- [x] #34403
- [x] #34408
- [x] #34619 

### 8. Core (core.py)

- [x] #34297
- [x] #34672 
- [x] #34693 
- [x] #34700 

### 9. Trainer (trainer.py)

- [x] #34419

## Contribution Guidelines

For fellow contributors, please follow this pattern.

Sample Issue: #29116  
Example PR: https://github.com/keras-team/keras/pull/20883  

Please group similar operations in the same issue and PR if they are implemented in the same file.

Previous description:
---

Hi team,

A first set of OpenVINO backend ops was implemented in keras-team/keras/issues/20910, which improved operator coverage.  
After reviewing the current state of `keras/backend/openvino` following that work, I identified another set of operations and utilities that are either unimplemented or could improve parity for this backend.

I’d like to propose the next structured batch of improvements here and get feedback from @rkazants before I start submitting PRs for them.

I would really appreciate guidance from the OpenVINO backend maintainers on:

- Which of these would be most valuable to prioritize
- Whether this aligns with the current priorities for the OpenVINO backend

**Ref:** keras-team/keras/issues/20910

I have some prior experience contributing to OpenVINO and am familiar with the OpenVINO opset operations, which should help in implementing these ops correctly.

### Issue submission checklist

- [x] The feature request or improvement must be related to OpenVINO

**Comment by rkazants** [comment]
Hi @goyaladitya05,

Please proceed with your plan. It will be great. If you feel to have a lack of time, try to share some ops with other contributors and collaborate with each other.

Best regards,
Roman

**Comment by goyaladitya05** [comment]
Thanks for the go-ahead.
I’ll start working on the first batch of ops and open PRs one by one, referencing this issue.

I’ll also split some of these into follow-up issues and mark a few of them as `good first issues` so other contributors can pick them up and collaborate.

**Comment by goyaladitya05** [comment]
Hi @mlukasze ,
I felt all these sub-issues were causing a lot of clutter and spam. So, instead I'll track everything in a main checklist here, directly linking PR's to this.
Is this approach alright, or do we need the sub-issues for milestoning? Which way should we proceed?

**Comment by daehyun99** [comment]
Hi @goyaladitya05 

It seems that `random.dropout` is not being tracked, and the task appears to be completed.

issue: https://github.com/openvinotoolkit/openvino/issues/34263
PR: https://github.com/keras-team/keras/pull/22263

**Comment by goyaladitya05** [comment]
Hey @andersendsa, I saw you've been contributing pretty actively to this. 

To keep things organized, we try to group similar functions and implementations within a single issue/PR if they belong in the same file. This helps in reducing clutter and spam.

Also, please ensure each PR has a description (even just 2-3 lines) explaining the changes. It really helps with the review process. 
Thanks for your contributions.

**Comment by daehyun99** [comment]
Hi there.

I opened the issue #34619
it is related `rnn.py`

Thanks.

**Comment by goyaladitya05** [comment]
Hi @daehyun99 @CityBoy-Claude,
Would either of you have time to work on the image ops?

**Comment by daehyun99** [comment]
Hi @goyaladitya05 
I can work part to part

**Comment by goyaladitya05** [comment]
> Hi [@goyaladitya05](https://github.com/goyaladitya05) I can work part to part

Sure, check out #34796. I initially opened a single issue for all 3 functions, but you can take it one at a time.

**Comment by daehyun99** [comment]
@goyaladitya05 
Oh, Someone already opened PR.
So, Can I work `extract_patches`?

**Comment by goyaladitya05** [comment]
> [@goyaladitya05](https://github.com/goyaladitya05) Oh, Someone already opened PR. So, Can I work `extract_patches`?

Yeah, please continue. You can take on gaussian_blur after that.

**Comment by goyaladitya05** [comment]
Hi @rkazants,
This roadmap is now complete. The OpenVINO backend has maximum possible functional parity compared to other backends. 
A total of  ~130 ops were implemented. Earlier approx ~1300 tests were in the excluded file (both concrete and directory level combined), which is now down to 17, after the merge of keras-team/keras/pull/22790. Please review, and let me know if any changes are required.

The following bugs were identified and reported -
1. #34780
2. #34638
3. #34138 
4. #35464

The following tests are still excluded due to the given reasons:
1. Explained in keras-team/keras/pull/22784
```
EqualizationTest::test_equalizes_to_all_bins
EqualizationTest::test_input_dtypes_float32
EqualizationTest::test_input_dtypes_int32
EqualizationTest::test_input_dtypes_int64
EqualizationTest::test_output_range_0_1
EqualizationTest::test_output_range_0_255
```

2. I could not find or implement proper and non buggy decompositions for these. Also, most of the linalg ops have used turnarounds due to openvinotoolkit/openvino/issues/35464. We can have another go at these once this bug is fixed upstream.
```
LinalgOpsCorrectnessTest::test_eig
LinalgOpsCorrectnessTest::test_lstsq
LinalgOpsCorrectnessTest::test_norm_2_-2
LinalgOpsCorrectnessTest::test_norm_2_2
LinalgOpsCorrectnessTest::test_norm_2_nuc
LinalgOpsCorrectnessTest::test_svd
```

3. openvinotoolkit/openvino/issues/34138 
```
MathOpsCorrectnessTest::test_erfinv_operation_basic
MathOpsCorrectnessTest::test_erfinv_operation_dtype
```

4. Could not find a way a decomposition for the beam_search path with the current opset. The greedy path has been implemented.
```
NNOpsCorrectnessTest::test_ctc_decode
NNOpsDtypeTest::test_ctc_decode
```

We have still kept the excluded_concrete_tests.txt file, since these above tests are excluded due to bugs.

**Comment by goyaladitya05** [comment]
while Keras has level parity, there needs to be done some more work on keras-hub, to align OpenVINO's implmentation on both, hence I'm keeping this issue open for now.

**Comment by goyaladitya05** [comment]
> while Keras has level parity, there needs to be done some more work on keras-hub, to align OpenVINO's implmentation on both, hence I'm keeping this issue open for now.

ref: keras-team/keras-hub/pull/2875

**Comment by goyaladitya05** [comment]
Hi, thank you everyone for your contributions. Since we have almost completed support, I'm closing this tracking issue now. The keras-hub work will be carried out seperately.
